aiokafka/message_accumulator.py

```diff
@@ -91,32 +98,40 @@ def append(self, key, value, timestamp_ms):
         msg += encoded
         self._buffer.write(msg)
 
-        future = asyncio.Future(loop=self._loop)
-        self._msg_futures.append(future)
         self._relative_offset += 1
-        return future
+        if future:
+            fut = asyncio.Future(loop=self._loop)
+            self._msg_futures.append(fut)
+            return fut
+        return self.future
 
     def done(self, base_offset=None, exception=None):
         """Resolve all pending futures"""
         for relative_offset, future in enumerate(self._msg_futures):
-            if future.done():
-                continue
-            if exception is not None:
-                future.set_exception(exception)
-            elif base_offset is None:
-                future.set_result(None)
-            else:
-                res = RecordMetadata(self._tp.topic, self._tp.partition,
-                                     self._tp, base_offset + relative_offset)
-                future.set_result(res)
-
-    def wait_deliver(self):
+            self._set_future(future, base_offset, relative_offset, exception)
+        # batch gets the offset of the first message in the set
+        self._set_future(self.future, base_offset, 0, exception)
+
+    def _set_future(self, future, base_offset, relative_offset, exception):
+        if future.done():
+            return
+        if exception is not None:
+            future.set_exception(exception)
+        elif base_offset is None:
+            future.set_result(None)
+        else:
+            res = RecordMetadata(self.tp.topic, self.tp.partition,
```

Comment:
I'm not sure how to treet RecordMetadata for batch resuts... Maybe a new class, like BatchMetadata...

Reply:
Whatever you'd prefer is fine by me. The offset strategy I chose here (returning the offset of the first message written) matches the value returned in ProduceResponse, so I think the values are sensible even if the class name is not.

Reply:
But `checksum`, `timestamp`, `serialized_key_size` and `serialized_value_size` don't really have any meaning thou.

Reply:
Add a new class to `structs.py`, say BatchMetadata. It will contain `offset`, `topic`, `partition` and `timestamp` fields, where the timestamp is the one from the server.

Reply:
> But checksum, timestamp, serialized_key_size and serialized_value_size don't really have any meaning thou.

Are you sure you're not looking at ConsumerRecord? RecordMetadata has none of those fields.

> Add a new class to structs.py, say BatchMetadata. It will contain offset, topic, partition and timestamp fields, where the timestamp is the one from the server.

This is almost identical to RecordMetadata (except for the timestamp field):
```python
RecordMetadata = collections.namedtuple(
    'RecordMetadata', ['topic', 'partition', 'topic_partition', 'offset'])
```

Like I said, I'm happy to add a new class if that's your preference, but I think we're looking at two different pieces of code right now.

Reply:
Ooops =) Looked at `kafka-python` implementation. Hmm, it's a bug actually, RecordBatch should have a `timestamp`. Will create an issue.